File: leetcode/word_ladder_II.py
import unittest

class Solution(object):
    def findLadders(self, beginWord, endWord, wordlist):
        """
        :type beginWord: str
        :type endWord: str
        :type wordlist: Set[str]
        :rtype: List[List[int]]
        """
        wordList = set(wordList)
        
    
class UnitTests(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.soln = Solution()

    def testIsNumber(self):
        beginWord = "hit"
        endWord = "cog"
        wordList = ["hot","dot","dog","lot","log"]
        result = [
            ["hit","hot","dot","dog","cog"],
            ["hit","hot","lot","log","cog"]
        ]
        
        self.assertEqual(self.soln.findLadders(beginWord, endWord, wordList), result)
        
if __name__ == "__main__":
    unittest.main()

File: leetcode_cpp/word_ladder_II.cpp
#include <vector>
#include <algorithm>
#include <unordered_set>

using namespace std;

class Solution {
public:

    vector<vector<string>> recursiveFindLadders(string beginWord, string endWord, unordered_set<string> &wordList, unordered_set<char>& unique_chars, vector<vector<string>>& ladders) {
        if (beginWord == endWord)
    }

    vector<vector<string>> findLadders(string beginWord, string endWord, unordered_set<string> &wordList) {
        vector<vector<string>> ladders;
        unordered_set<char> unique_chars;
        
        // Add all the unique characters.
        for (auto word : wordList) {
            for (auto c : word) {
                unique_chars.insert(c);
            }
        }

        recursiveFindLadders(beginWord, endWord, wordList, unique_chars, ladders);
        return ladders;
    }
};

int main() {
    Solution soln;
    string beginWord = "hit";
    string endWord = "cog";
    unordered_set<string> wordList { "hot","dot","dog","lot","log" };
    vector<vector<string>> ladders = soln.findLadders(beginWord, endWord, wordList);
}